core/src/main/java/org/geysermc/geyser/item/GeyserCustomItemManager.java

```diff
@@ -0,0 +1,181 @@
+import com.nukkitx.protocol.bedrock.packet.StartGamePacket;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.geysermc.geyser.GeyserImpl;
+import org.geysermc.geyser.api.item.custom.CustomItemData;
+import org.geysermc.geyser.api.item.custom.CustomItemManager;
+import org.geysermc.geyser.api.item.custom.CustomItemOptions;
+import org.geysermc.geyser.api.item.custom.NonVanillaCustomItemData;
+import org.geysermc.geyser.item.mappings.MappingsConfigReader;
+import org.geysermc.geyser.registry.populator.CustomItemRegistryPopulator;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GeyserCustomItemManager extends CustomItemManager {
+    public static final String CUSTOM_PREFIX = "geyser_custom:";
+
+    private Map<String, List<CustomItemData>> customMappings = new HashMap<>();
+
+    private List<GeyserCustomMappingData> registeredCustomItems = new ArrayList<>();
+    private Int2ObjectMap<List<ComponentItemData>> registeredComponentItems = new Int2ObjectOpenHashMap<>();
+
+    public GeyserCustomItemManager() {
+        MappingsConfigReader.init(this);
+    }
+
+    public void loadMappingsFromJson() {
+        Path customMappingsDirectory = MappingsConfigReader.getCustomMappingsDirectory();
+        if (!Files.exists(customMappingsDirectory)) {
+            try {
+                Files.createDirectories(customMappingsDirectory);
+            } catch (IOException e) {
+                GeyserImpl.getInstance().getLogger().error("Failed to create custom mappings directory", e);
+                return;
+            }
+        }
+
+        Path[] mappingsFiles = MappingsConfigReader.getCustomMappingsFiles();
+        for (Path mappingsFile : mappingsFiles) {
+            this.loadMappingsFromJson(mappingsFile);
+        }
+
+        if (this.registeredItemCount() != 0) {
+            GeyserImpl.getInstance().getLogger().info("Registered " + this.registeredItemCount() + " custom items from mappings");
+        }
+    }
+
+    public void loadMappingsFromJson(@NotNull Path file) {
+        MappingsConfigReader.readMappingsFromJson(file);
+    }
+
+    private boolean cantRegisterNewItem() {
+        return GeyserImpl.getInstance().isInitialized() || !GeyserImpl.getInstance().getConfig().isAddNonBedrockItems();
+    }
+
+    @Override
+    public boolean registerCustomItem(@NotNull String identifier, @NotNull CustomItemData customItemData) {
+        if (this.cantRegisterNewItem()) {
+            return false;
+        }
+
+        GeyserCustomMappingData registeredItem = CustomItemRegistryPopulator.populateRegistry(identifier, customItemData, this.registeredItemCount());
```

Comment:
Why is the whole registry repopulated every time we register an item.
If we have to do this. Isn't this something we can do once, once all the custom items have been registered?

Reply:
I'm not sure what you mean? It just adds an item to the registry. It's not the whole thing.